cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java

```diff
@@ -0,0 +1,184 @@
+    // Delete all apps
+    appClient.deleteAll(Id.Namespace.DEFAULT);
+  }
+
+  private Map<String, String> deployAppVersionStartService(ApplicationId app) throws Exception {
+    // Deploy app of given version
+    AppRequest createRequest = new AppRequest<>(
+      new ArtifactSummary(artifactId.getArtifact(), artifactId.getVersion()));
+    appClient.deploy(app, createRequest);
+    // Start the service of the app
+    ServiceId service = app.service(AppReturnsArgs.SERVICE);
+    Map<String, String> propMap = Maps.newHashMap();
+    propMap.put("version", app.getVersion());
+    propMap.put("logical.start.time", "1234567890000");
+    programClient.start(service, false, propMap);
+    assertProgramRunning(programClient, service);
+    return propMap;
+  }
+
+  private void addAppReturnsArgsArtifact() throws Exception {
+    LocalLocationFactory locationFactory = new LocalLocationFactory(TMP_FOLDER.newFolder());
+    Manifest manifest = new Manifest();
+    manifest.getMainAttributes().put(ManifestFields.BUNDLE_VERSION, "1.0.0");
+    final Location appJarLoc = AppJarHelper.createDeploymentJar(locationFactory, AppReturnsArgs.class, manifest);
+    InputSupplier<InputStream> inputSupplier = new InputSupplier<InputStream>() {
+      @Override
+      public InputStream getInput() throws IOException {
+        return appJarLoc.getInputStream();
+      }
+    };
+    artifactClient.add(artifactId.getParent().toId(), artifactId.getArtifact(), inputSupplier, artifactId.getVersion());
+    appJarLoc.delete();
+  }
+
+  private void assertArgsEqual(Map<String, String> expectedArgs)
+    throws UnauthorizedException, IOException, UnauthenticatedException {
+    HttpResponse response = routeConfigClient.callServiceMethod(NamespaceId.DEFAULT, AppReturnsArgs.NAME,
+                                                                AppReturnsArgs.SERVICE, AppReturnsArgs.ENDPOINT);
+    Map<String, String> responseMap = GSON.fromJson(response.getResponseBodyAsString(), STRING_MAP_TYPE);
+    Assert.assertEquals(expectedArgs, responseMap);
+  }
+
+  private Map<String, Integer> getRouteConfig() throws IOException, UnauthorizedException, UnauthenticatedException {
+    return routeConfigClient.getRouteConfig(NamespaceId.DEFAULT, AppReturnsArgs.NAME, AppReturnsArgs.SERVICE);
+  }
+
+  private void storeRouteConfig(Map<String, Integer> routeConfig)
+    throws UnauthorizedException, IOException, UnauthenticatedException {
+    routeConfigClient.storeRouteConfig(NamespaceId.DEFAULT, AppReturnsArgs.NAME, AppReturnsArgs.SERVICE, routeConfig);
+  }
+
+  private void storeAndGetValidRouteConfig(Map<String, Integer> routeConfig) throws Exception {
+    storeRouteConfig(routeConfig);
+    Assert.assertEquals(routeConfig, getRouteConfig());
+  }
+
+  private void storeInvalidRouteConfig(Map<String, Integer> routeConfig, String expectedMsg) throws Exception {
+    try {
+      storeRouteConfig(routeConfig);
+    } catch (IOException expected) {
```

Comment:
`RouteConfigHttpHandler` throws BadRequestException but `RouteConfigClient` can only get IOException from the HTTP response. BadRequestException can be read from the HTTP status code from the response message.